src/runtime/HalideBuffer.h

```diff
@@ -1069,11 +1091,25 @@ class Buffer {
     }
     // @}
 
+#if defined(HALIDE_RUNTIME_BUFFER_WRAPPERS)
+    // All Buffer<T> can produce halide_typed_buffer_t<const T>
+    operator halide_typed_buffer_t<typename std::add_const<T>::type> () {
+        return halide_typed_buffer_t<typename std::add_const<T>::type>(raw_buffer());
+    }
+
+    // Only Buffer<non-const-T> can produce halide_typed_buffer_t<non-const-T>
+    operator halide_typed_buffer_t<typename std::remove_const<T>::type> () {
+        static_assert(!std::is_const<T>::value, "Can't convert from a Buffer<const T> to a Buffer<T>");
+        return halide_typed_buffer_t<typename std::remove_const<T>::type>(raw_buffer());
+    }
+
+#else
     /** Provide a cast operator to halide_buffer_t *, so that
      * instances can be passed directly to Halide filters. */
     operator halide_buffer_t *() {
```

Comment:
Per comment from @abadams in https://github.com/halide/Halide/issues/3963, should we mark `operator halide_buffer_t *` with HALIDE_ATTRIBUTE_DEPRECATED here as well?

Reply:
Probably should port over all the production code we're aware of first, in case it takes a while to get done.